### PR TITLE
Plugin - "preformatted" able to use pre/code tags automatically

### PR DESCRIPTION
**Pull request by Civile**
Hi, i've added a sketch of the preformatted plugin. Now it basically works but i think that in the future I will continue to make it more functional and specific (like the code-preview tool of the StackOverflow wysiwyg or something like that, cause i find it very useful). 

When you'll have time can you review it please? Thanks Alex

### CODE REVIEW

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                preformatted: "Exemple de code",
+            },
+            it: {
+              preformatted: "Codice <pre>",
+            }
+        },
+        opts: {
+            btnsDef: {
+                preformatted: {
+                    func: function(params, btw) {
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: It's `tbw` (for **T**rum**b**o**w**yg) not `btw`

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                        var t = btw;
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: You can remove that :)

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                        var text = String(t.doc.getSelection());
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: And use `tbw` insteed of `t`

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                        if(text.replace(/\s/g, '') !== '') {
+                            try {
+                                var curtag = getSelectionParentElement().tagName.toLowerCase();
+                                if(curtag == "code" || curtag == "pre") {
+                                    return unwrapCode();
+                                }
+                                else {
+                                    btw.execCmd('insertHTML', "<pre><code>"+ strip(text) +"</code></pre>");
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: Here it's `tbw`

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                }
+            }
+        },
+    });
+    
+    /*
+     * GetSelectionParentElement
+    */
+    function getSelectionParentElement() {
+        var parentEl = null, sel;
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: I prefer like that:

```
var parentEl = null,
    sel;
```

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            $(container).contents().unwrap('code').unwrap("pre");
+        } else if(ispre) {
+            $(container).contents().unwrap("pre");
+        } else if(iscode) {
+            $(container).contents().unwrap("code");
+        }  
+    }
+
+
+})(jQuery);
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: Use single quotes to keep code homogeneous

File: plugins/preformatted/ui/sass/trumbowyg.preformatted.scss [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+@import "sprite-black-2x";
+@import "sprite-white";
+@import "sprite-white-2x";
+
+@import "sprite-pos";
+
+$preformatted-black: 0px 0px;
+$preformatted-white: 0px 0px;
+$preformatted-black-2x: 0px 0px;
+$preformatted-white-2x: 0px 0px;
```
Comment (Alex-D) [inline plugins/preformatted/ui/sass/trumbowyg.preformatted.scss]: These 4 lines are useless, they are generated in 

```
@import "sprite-black";
@import "sprite-black-2x";
@import "sprite-white";
@import "sprite-white-2x";
```

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    $.extend(true, $.trumbowyg, {
+        langs: {
+            en: {
+                preformatted: "Code sample <pre>",
+            },
+            fr: {
+                preformatted: "Exemple de code",
+            },
+            it: {
+              preformatted: "Codice <pre>",
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: 2 spaces more to get 4 spaces :)

File: plugins/preformatted/trumbowyg.preformatted.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        var container = null;
+        if (document.selection) //for IE
+            container = document.selection.createRange().parentElement();
+        else {
+            var select = window.getSelection();
+            if (select.rangeCount > 0)
+                container = select.getRangeAt(0).startContainer.parentNode;
+        }
+        //"paranoic" unwrap
+        var ispre  = $(container).contents().closest("pre").length;
```
Comment (Alex-D) [inline plugins/preformatted/trumbowyg.preformatted.js]: Remove the useless space before `=`

**Comment by Alex-D** [comment]
That's sounds great! Maybe you can add the isSupported method to check if browser support exeCommand insertHTML like here : https://github.com/Alex-D/Trumbowyg/blob/master/plugins/base64/trumbowyg.base64.js#L32

**Comment by Civile** [comment]
Yes sure i will add that too. Thanks Alex, i'll update the plugin as soon as i made the corrections that you have suggested ;)

File: plugins/preformatted/trumbowyg.preformatted.js (hunk 6 above)
Comment (Civile) [inline plugins/preformatted/trumbowyg.preformatted.js]: Here, should i use it on all the entire document? Single quotes instead of double

**Comment by Alex-D** [comment]
Mhhh ok! Merged :)

**Comment by Civile** [comment]
Great! Here we start our collaboration! Thank you Alex ;)

**Comment by navneet35371** [comment]
(y)

**Pull request merged**